Auto save published post upon preview and show preview url

yaelirub opened the pull request:
Fixes #9287  (also refer to #9282 )
[autosave API call PR on WordPressKit](https://github.com/wordpress-mobile/WordPressKit-iOS/pull/155)

The gist:
When asking to preview a post we check if the post has unsaved changes (now also returning true if the post status is the new status: `AbstractPostRemoteStatusSaved`). 
If the post is not draft, we call the save endpoint that checks if the post blog supports WPRestAPI since this endpoint is not available for self hosted sites.
Not WPCom - we return failure with PostServiceErrorDomain error. and show a **noResultsViewController with "preview not available" - needs copy/ design approval (please see gif)**
Is WPCom - we call the autosave endpoint and get the previewURL on successful response.
We pass the previewURL to the preview generator to display.

WPCom site preview published post:

![preview1](https://user-images.githubusercontent.com/1335657/57185878-3cbb6600-6e89-11e9-8e14-342b8e44c393.gif)

Self hosted site preview published post:
![IMG_4064](https://user-images.githubusercontent.com/1335657/57185884-5066cc80-6e89-11e9-87cd-30dfd173d0a7.GIF)

To test:
 **Following code review and testing by developers, we have discovered an issue where the preview doesn't show the changes for a Jetpack site.**
A. preview post no updates 

1. go to a published post from your post list.
2. go to preview.
3. see the post preview with the latest content.

B. Preview post with updates
1. go to a published post from your post list
2. edit the post
3. preview the post
4. see "Generating preview..."
5. see the post preview with the latest content.

C. preview after discard
1. Steps in B
2. Exit the editor
2. Tap "Discard" when prompted in the action sheet
3. go back to the post
4. hit preview 
5. see the post preview with the latest (published) content.

* I will update the release notes once approved.
- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

aerych commented on WordPress/Classes/Services/PostService.h:
```diff
@@ -92,6 +92,21 @@ extern const NSUInteger PostServiceDefaultNumberToSync;
+ succeeds.
+ @param failure A failure block
+ */
+- (void)savePost:(AbstractPost *)post
```
Same question about naming that I had in https://github.com/wordpress-mobile/WordPressKit-iOS/pull/156

aerych commented on WordPress/Classes/Services/PostService.m:
```diff
@@ -271,6 +270,62 @@ - (void)uploadPost:(AbstractPost *)post
+            AbstractPost *postInContext = (AbstractPost *)[self.managedObjectContext existingObjectWithID:postObjectID error:nil];
+            if (postInContext) {
+                postInContext.remoteStatus = AbstractPostRemoteStatusSaved;
+                NSPredicate *unattachedMediaPredicate = [NSPredicate predicateWithFormat:@"postID <= 0"];
```
The unattached media handling is almost identical between the `savePost` and `updatePost` methods, the only difference is the parameters in the success blocks.   I wonder if we could refactor this to its own method to avoid the duplication?

aerych commented on WordPress/Classes/ViewRelated/Post/PostEditor+MoreOptions.swift:
```diff
@@ -23,33 +23,68 @@ extension PostEditor where Self: UIViewController {
+        let publishedStatus = NSLocalizedString("Generating Preview...", comment: "Text displayed in HUD while a post is being saved.")
+        SVProgressHUD.setDefaultMaskType(.clear)
+
+        if post.hasUnsavedChanges() {
```
Nitpicky, but maybe this could be tweaked to avoid the nested conditionals?  Something like
```
if !post.hasUnsavedChanges() {
      completion(nil, nil)
      return
}
```

aerych commented:
Tested as described and worked for me!  I also tried with a post with the pending review status. No problems.

yaelirub commented:
@aerych , thank you so much for your review. I updated the PR based on your comments. Ready for another look.

yaelirub commented:
@megsfulton , moving towards removing "fake previews", we are now showing this screen when the user asks to preview a post from a self hosted site. Please your advice 🙏 
![Simulator Screen Shot - iPhone Xs - 2019-05-08 at 21 59 59](https://user-images.githubusercontent.com/1335657/57428317-a8306b00-71dc-11e9-925a-1c274e10eaec.png)

aerych commented:
So... I'm not sure what to make of this.   When I was initially testing a self-hosted Jetpack connected site, I was not seeing a preview that included the changes. 

### Example 1 - Changes missing
![autosave 2019-05-09 10_41_32](https://user-images.githubusercontent.com/1435271/57476434-6b4f8d00-724b-11e9-92f4-a6f151430614.gif)
Note that the site here is a jurassic.ninja test site but its showing the same behavior that I saw on my own test site -- demonstrating that the behavior should be reproducible. 

But as I continue to test I do see that changes are showing up when previewing. 

### Example 2 - Changes present
![autosave 2 2019-05-09 11_11_44](https://user-images.githubusercontent.com/1435271/57476494-8e7a3c80-724b-11e9-92f4-089cb9aab966.gif)

Example 2 shows my test site but now previews are working.  

I am quite perplexed.

megsfulton commented:
>moving towards removing "local previews", we are now showing this screen when the user asks to preview a post from a self hosted site. 


Some copy suggestions here to make it clear why the preview is unavailable: 

Title:
**Preview Unavailable for Self-Hosted Sites** 

Subtitle:
1) The updates to your post need to be published to be viewed.
or
2) To view the changes, publish the updates to your post. 

-

One other thing, in the gif you shared for "WPCom site preview published post:" -- I noticed "Privately Published" in the post list -- Just want to check that the "Privately Published" label there is unrelated to the preview changes in this issue and it's showing because the post you were updating was a private post?

diegoreymendez commented:
> Note that the site here is a jurassic.ninja test site but its showing the same behavior that I saw on my own test site -- demonstrating that the behavior should be reproducible.

But as I continue to test I do see that changes are showing up when previewing.

@aerych - Not sure if this is relevant but I recall seeing the same behavior using the web editor, where I had to preview twice for the autosave to be properly loaded.

yaelirub commented:
>Or if we want to be paranoid .jp sites could be treated like self-hosted sites.
I'm testing some more but think that might be what I'd end up doing.

yaelirub commented:
@megsfulton , thanks so much for your input. I have updated the view as suggested:
![Simulator Screen Shot - iPhone Xs - 2019-05-09 at 14 52 38](https://user-images.githubusercontent.com/1335657/57489947-8383d400-726c-11e9-98ed-b0671d6e3cf1.png)

> One other thing, in the gif you shared for "WPCom site preview published post:" -- I noticed "Privately Published" in the post list -- Just want to check that the "Privately Published" label there is unrelated to the preview changes in this issue and it's showing because the post you were updating was a private post?

The privately published label is unrelated. I set it as a private post.

aerych commented:
> I recall seeing the same behavior using the web editor, where I had to preview twice for the autosave to be properly loaded.

Potentially relevant.  Was it when previewing a jetpack site via the wpcom rest api by any chance? 

> I'm testing some more but think that might be what I'd end up doing.

I'm torn about this but I also don't have another answer.

diegoreymendez commented:
> Potentially relevant. Was it when previewing a jetpack site via the wpcom rest api by any chance?

In my case it was one of my WP.com test sites.

My feeling around this is that it's a problem with the autosaves API, or something similar... one possibility could be to record this as a known issue and split it from this PR.

It is my opinion that this PR already offers a much better experience than what we had before, even with this bug.

diegoreymendez commented on WordPress/Classes/Models/AbstractPost.h:
```diff
@@ -13,6 +13,7 @@ typedef NS_ENUM(NSUInteger, AbstractPostRemoteStatus) {
     AbstractPostRemoteStatusLocal,       // Only local version
     AbstractPostRemoteStatusSync,       // Post uploaded
     AbstractPostRemoteStatusPushingMedia, // Push Media
+    AbstractPostRemoteStatusSaved,       // Post saved
```
If this is specifically refering to autosaved posts, can we rename it to something like `AbstractPostRemoteStatusAutoSaved`?

It's a minor difference, but "saving" can a bit vague and be confused with the regular uploading of the post, whereas "autosaving" has a more concrete meaning (at least in my head).

diegoreymendez replied:
Also, if possible, update the comment to be "Post auto-saved" as well.

diegoreymendez commented on WordPress/Classes/Services/PostService+UnattachedMedia.swift:
```diff
@@ -0,0 +1,22 @@
+            ContextManager.sharedInstance().save(self.managedObjectContext)
+            success()
+        }) { error in
+            success()
```
Can you provide some background regarding why on failure we are reporting success?

The reason I'm asking is because if we decide we want to mute a specific error, it's usually best to leave that decision for the code that's calling this method, rather than to have this method automatically silence the error.

yaelirub replied:
I deleted the comment that explained this when I refactored the method:

'// Sergio Estevao (2018-02-27): even if media fails to attach we are answering with success because the post uploaded successful and the only thing that failed was attaching the media to it.'
I updated the callback to `completion` instead of `success` as Eric suggested but realize that you are right and it would be better handled by the caller so I added a failure block, logging the failure and added back the comment.

diegoreymendez commented on WordPress/Classes/Services/PostService.h:
```diff
@@ -92,6 +92,21 @@ extern const NSUInteger PostServiceDefaultNumberToSync;
+ */
+- (void)autoSave:(AbstractPost *)post
+         success:(nullable void (^)(AbstractPost *post, NSString *previewURL))success
+           failure:(void (^)(NSError * _Nullable error))failure;
```
Nitpicky: there seems to be an indentation issue with this line.

yaelirub replied:
Great catch!

megsfulton commented:
Looking at this with fresh eyes today, I think this copy might be clearer for that screen:

Preview Unavailable for Published Posts
Update the published post to view your changes.

yaelirub commented:
Thanks @megsfulton , updated:
![Simulator Screen Shot - iPhone Xs - 2019-05-10 at 16 10 12](https://user-images.githubusercontent.com/1335657/57560996-3839f600-733e-11e9-858e-b49b682fdb4e.png)

yaelirub commented:
@diegoreymendez, updated the code according to your comments. 
>My feeling around this is that it's a problem with the autosaves API, or something similar... one possibility could be to record this as a known issue and split it from this PR.

I can create a ticket for this. Do you agree, @aerych ?

aerych commented:
> I can create a ticket for this

Sounds reasonable.  :+1:

aerych commented:
Hmm. I marked approved but maybe I should have dismissed my earlier review instead? 🤔 
Regardless, its 👍 from me with Diego's recommended changes + the conflicts resolved.
